File: rules-interpreter/src/main/java/io/github/iamrenny/ruleflow/evaluators/ValidPropertyContextEvaluator.java
package io.github.iamrenny.ruleflow.evaluators;

import io.github.iamrenny.ruleflow.RuleFlowLanguageParser.ValidPropertyContext;
import io.github.iamrenny.ruleflow.errors.PropertyNotFoundException;
import io.github.iamrenny.ruleflow.visitors.Visitor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

public class ValidPropertyContextEvaluator implements ContextEvaluator<ValidPropertyContext> {
    private static final Logger logger = LoggerFactory.getLogger(ValidPropertyContextEvaluator.class);

    @Override
    public Object evaluate(ValidPropertyContext ctx, Visitor visitor) throws PropertyNotFoundException {
        String property = ctx.getText();
        Object result = visitor.getData().get(property);
        logger.debug("ValidProperty: property={}, result={}", property, result);
        if (ctx.property != null) {
            if (result == null) {
                throw new PropertyNotFoundException(ctx.ID(0).getText() + " field cannot be found");
            }
            return result;
        } else if (ctx.nestedProperty != null) {
            Map<String, ?> data = (ctx.root != null) ? visitor.getRoot() : visitor.getData();
            return getNestedValue(ctx, data);
        } else {
            throw new PropertyNotFoundException(ctx.getText() + " field cannot be found");
        }
    }

    @SuppressWarnings("unchecked")
    private Object getNestedValue(ValidPropertyContext ctx, Map<String, ?> data) throws PropertyNotFoundException {
        Map<String, ?> currentData = data;
        for (var id : ctx.ID()) {
            Object value = currentData.get(id.getText());
            if (value instanceof Map<?, ?>) {
                currentData = (Map<String, ?>) value;
            } else {
                if (value == null) {
                    throw new PropertyNotFoundException(id.getText() + " field cannot be found");
                }
                return value;
            }
        }
        throw new PropertyNotFoundException(ctx.getText() + " cannot be found");
    }
}

File: rules-interpreter/src/main/kotlin/com/rappi/analang/evaluators/ValidPropertyContextEvaluator.kt
package com.rappi.analang.evaluators

import com.rappi.fraud.analang.ANAParser
import com.rappi.analang.errors.PropertyNotFoundException
import com.rappi.analang.visitors.Visitor

class ValidPropertyContextEvaluator: ContextEvaluator<ANAParser.ValidPropertyContext> {
    override fun evaluate(ctx: ANAParser.ValidPropertyContext, visitor: Visitor): Any {
        return when {
            ctx.property != null -> visitor.data[ctx.ID(0).text]
                ?: throw PropertyNotFoundException("${ctx.ID(0).text} field cannot be found")
            ctx.nestedProperty != null -> {
                val data = if(ctx.root != null) {
                    visitor.root
                } else {
                    visitor.data
                }
                getNestedValue(
                    ctx,
                    data
                )
            }

            else -> throw PropertyNotFoundException("${ctx.text} field cannot be found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun getNestedValue(ctx: ANAParser.ValidPropertyContext, data: Map<String, *>): Any {
        var r = data
        ctx.ID().forEach { id ->
            when {
                r[id.text] is Map<*, *> -> r = r[id.text] as Map<String, Any?>
                else -> return r[id.text] ?: throw PropertyNotFoundException("${id.text} field cannot be found")
            }
        }
        throw PropertyNotFoundException("${ctx.text} cannot be found")
    }
}